hal-x86_64/src/interrupt/apic/io.rs

```diff
@@ -81,11 +91,184 @@ struct MmioRegisters {
+
+        // There may be multiple I/O APICs in the system, so we'll collect them
+        // all into a `Vec`. We're also going to build a table of how the ISA
+        // standard interrupts are mapped across those I/O APICs, so we can look
+        // up which one a particular interrupt lives on.
+        let n_ioapics = madt.io_apics.len();
+        tracing::trace!(?madt.io_apics, "found {n_ioapics} IO APICs", );
+        assert_ne!(
+            n_ioapics, 0,
+            "why would a computer have the APIC interrupt model but not \
+             have any IO APICs???"
+        );
+        let mut this = IoApicSet {
+            ioapics: alloc::vec::Vec::with_capacity(n_ioapics),
+            isa_map: [IsaOverride { apic: 0, vec: 0 }; 16],
+        };
+
+        for (n, ioapic) in madt.io_apics.iter().enumerate() {
+            let addr = PAddr::from_u64(ioapic.address as u64);
+            tracing::debug!(ioapic.paddr = ?addr, "IOAPIC {n}");
+            this.ioapics
+                .push(Mutex::new(IoApic::new(addr, pagectrl, frame_alloc)));
+        }
+
+        // Okay, so here's where it gets ~*weird*~.
+        //
+        // On the Platonic Ideal Normal Computer, the ISA PC interrupts would be
+        // mapped to I/O APIC input pins by number, so ISA IRQ 0 would go to pin
+        // 0 on I/O APIC 0, and so on.
+        //
+        // However, motherboard manufacturers can do whatever they like when
+        // routing these interrupts, so they could go to different pins,
+        // potentially on different I/O APICs (if the system has more than one).
+        // Also, some of these interrupts might have different pin polarity or
+        // edge/level-triggered-iness than what we might expect.
+        //
+        // Fortunately, ACPI is here to help (statements dreamed up by the
+        // utterly deranged).
+        //
+        // The MADT includes a list of "interrupt source overrides" that
+        // describe any ISA interrupts that are not mapped to I/O APIC pins in
+        // numeric order. Each entry in the interrupt source overrides list will
+        // contain:
+        // - an ISA IRQ number (naturally),
+        // - the "global system interrupt", which is NOT the IDT vector for that
+        //   interrupt but instead the I/O APIC input pin number that the IRQ is
+        //   routed to,
+        // - the polarity and trigger mode of the interrupt pin, if these are
+        //   different from the default bus trigger mode.
+        // We can walk this
+        for irq in IsaInterrupt::ALL {
+            // Assume the IRQ is mapped to the I/O APIC pin corresponding to
+            // that ISA IRQ number, and is active-high and edge-triggered.
+            let mut global_system_interrupt = irq as u8;
+            let mut polarity = PinPolarity::High;
+            let mut trigger = TriggerMode::Edge;
+            // Is there an override for this IRQ? If there is, clobber the
```

Comment:
i'll admit that part of my thinking was hoping this didn't need to have the shape
```
// set some defaults
//
// put the defaults through an extruder that maybe messes with them
//
// now use whatever's left
```
or at least could do so in a more structured way, but with there being a SameAsBus polarity idk